### ISSUE TITLE
A bug in FileStream (async + buffer) since .Net 6?

### ISSUE THREAD

**Issue by yeunglee** [dotnet/runtime#80344]
### Description

I encounter a strange issue in FileStream. I provide the code and test data in https://github.com/yeunglee/file_stream_test

The major issue is that (async random read + special buffer size + special file) in FileStream encounters unexpected read.

Here is a simple code. The full code can be retrieved from https://github.com/yeunglee/file_stream_test
```
const string FILE_PATH = "test_data";

// SOLUTION #1: change buffer size. for example: 1024, 4096, 1024 * 512 - 1, 1024 * 512 + 1
const int FILE_BUFFER_SIZE = 1024 * 512;

Console.WriteLine($".Net version: {Environment.Version}");
Console.WriteLine($"OS version: {Environment.OSVersion}");

var fileData = await File.ReadAllBytesAsync(FILE_PATH);
// SOLUTION #2: add FileOptions.RandomAccess
await using var fs = new FileStream(FILE_PATH, FileMode.Open, FileAccess.Read, FileShare.ReadWrite, FILE_BUFFER_SIZE, FileOptions.Asynchronous);

void SetPosition(long p) => fs.Position = p;

async Task ReadAsync(int size)
{
    var initialPosition = fs.Position;
    var content = new byte[size];

    // SOLUTION #3: use sync Read instead of async ReadAsync
    var count = await fs.ReadAsync(content, 0, content.Length);

    var isMatched = fileData.Skip((int)initialPosition).Take(count).SequenceEqual(content.Take(count));

    if (!isMatched)
    {
        Console.WriteLine($"!!!Content is inconsistent!!! position={initialPosition:#,0} size={size:#,0} count={count:#,0}");
        Environment.Exit(1);
    }
}

await ReadAsync(1);
await ReadAsync(4);
await ReadAsync(4);
await ReadAsync(118);
await ReadAsync(4);
await ReadAsync(524288);
SetPosition(524377);
await ReadAsync(524288);
SetPosition(1048633);
await ReadAsync(524288);
SetPosition(1572864);
await ReadAsync(524288);
await ReadAsync(524288);
SetPosition(2621431);
await ReadAsync(524288);
```

### Reproduction Steps

1) Checkout the demo code from https://github.com/yeunglee/file_stream_test
2) Execute `dotnet run`  (The dotnet project file specify .Net 6, you can change .Net version)
3) Check the execution result from step#2.

### Expected behavior

The demo program outputs:
`FileStream is good`

### Actual behavior

The demo program outputs:
`!!!Content is inconsistent!!! position=2,621,431 size=524,288 count=326,069`

### Regression?

.Net 3.1 and .Net 5 are fine.

### Known Workarounds

There are ~~three~~ four workarounds:
1) Change file buffer size
2) Or change file option (add FileOptions.RandomAccess)
3) Or use sync Read instead of async ReadAsync

The workaround ways are commented in demo program code.

4) The demo code is fine with .Net 5, .Net 3.1

### Configuration

This issue can be reproduced in

1) WinServer 2022 (21H2) (x64)+ .Net 6.0.3
2) Win11 Enterprise (22H2) (x64) + .Net 7.0.1
3) Win11 Enterprise (22H2) (x64) + .Net 6.0.12
4) Win10 LTSC (1809) (x64) + .Net 6.0.11

Both Debug and Release configuration are same result.

**Comment by KeterSCP** [comment]
From the **Remarks** section of https://learn.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=net-7.0

> An implementation is free to return fewer bytes than requested even if the end of the stream has not been reached.

To correctly read all requested bytes, you need to change this line of code

```c#
var count = await fs.ReadAsync(content, 0, content.Length);
```

to

```c#
var count = 0;
while (count < content.Length)
{
    int bytesRead = fs.Read(content.AsSpan()[count..]);
    if (bytesRead == 0) break;
    count += bytesRead;
}
```

**Comment by yeunglee** [comment]
@adamsitnik 

# CAUSE
This bug is caused by
https://github.com/dotnet/runtime/blob/dcda1e08cb8f421877388be2d685c009a31b1845/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs#L319-L324

And
https://github.com/dotnet/runtime/blob/dcda1e08cb8f421877388be2d685c009a31b1845/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs#L914-L935


# FIX
The fix is pretty easy. Just make the read buffer unavailable. Like what it does:
https://github.com/dotnet/runtime/blob/dcda1e08cb8f421877388be2d685c009a31b1845/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs#L426-L439

e.g.  Add `_readLen = _readPos = 0;` before `return _strategy.ReadAsync(buffer, cancellationToken);` (Lines 319 to 324)

# QUESITON
But I don't understand the comment:
> To find out why we are bypassing cache here, please see WriteAsync comments.

Any idea?

**Comment by adamsitnik** [comment]
@yeunglee big thanks for a very detailed bug report, I've sent a fix in #80432

> But I don't understand the comment:

The comment refers to the following comment, which explains why we have such perf optimization:

https://github.com/dotnet/runtime/blob/7a6c31c107e4da7588f1dca65d4fbe65f52339f9/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs#L617-L630

**Comment by yeunglee** [comment]
@adamsitnik 

Sorry to bother you. I check your fix https://github.com/dotnet/runtime/pull/80432

It still has flaw. Reason is that
The hot path #1 condition is `if (_readLen == _readPos && buffer.Length >= _bufferSize)`

In new fix, it should be
(A) `if (_readLen == _readPos && _writePos == 0 && buffer.Length >= _bufferSize)` as hot path#1  new condition
OR
(B) *TRY TO WRITE FLUSH* in hot path#1

I prefer the option (A)

**Comment by yeunglee** [comment]
> @yeunglee big thanks for a very detailed bug report, I've sent a fix in #80432
> 
> > But I don't understand the comment:
> 
> The comment refers to the following comment, which explains why we have such perf optimization:
> 
> https://github.com/dotnet/runtime/blob/7a6c31c107e4da7588f1dca65d4fbe65f52339f9/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/BufferedFileStreamStrategy.cs#L617-L630

You're welcome. I posted a wrong comment. I have already deleted it.
